Configurable Private Cluster Dns/Fqdn

Gordonby opened the pull request:
## PR Summary

Introduces 2 new parameters to provide more control over Private Cluster DNS
1. privateClusterDnsMethod
2. dnsApiPrivateZoneId

Closes #235 (Private Cluster Fqdn)
Closes #328 (K8S Version)

### New default behaviour for private clusters

![image](https://user-images.githubusercontent.com/17914476/176390120-4e989e04-987f-48ea-98f9-edc77a87ae31.png)

mc_ resource group
![image](https://user-images.githubusercontent.com/17914476/176392786-59382a67-30fc-4931-819a-32f596d62a15.png)

### Behaviour for custom dns clusters

![image](https://user-images.githubusercontent.com/17914476/176433138-90ab211a-198f-4929-a8a4-3b52677148f6.png)

![image](https://user-images.githubusercontent.com/17914476/176433283-5c7c4130-ca5e-4055-ac94-4f58d41424c7.png)


### UI Changes

![image](https://user-images.githubusercontent.com/17914476/176426585-49726d79-b42c-490f-b4ba-d249dffd65ea.png)


### TODO

- [x] Discuss new parameters and logic
- [x] Build bicep code
- [x] Make UI changes
- [x] UI : Add validation of value (regex) for Dns Zone Id
- [x] UI : Adjust preset values for existing preset scenarios
- [x] Test and screengrab different deployment configurations

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] This PR is ready to merge and is not **Work in Progress**
- [x] Link to a filed issue
- [x] Screenshot of UI changes (if PR includes UI changes)

khowling commented:
Doesn't look like the bicep is assigning the agentpool identity roles to my dnszone, here is the paramters

```
az deployment group create -g az-k8s-3xt4-rg  --template-file ./bicep/main.bicep --parameters \
	resourceName=az-k8s-3xt4 \
	agentCount=2 \
	upgradeChannel=stable \
	JustUseSystemPool=true \
	custom_vnet=true \
	CreateNetworkSecurityGroups=true \
	enable_aad=true \
	AksDisableLocalAccounts=true \
	enableAzureRBAC=true \
	adminPrincipalId=$(az ad signed-in-user show --query id --out tsv) \
	registries_sku=Basic \
	acrPushRolePrincipalId=$(az ad signed-in-user show --query id --out tsv) \
	azurepolicy=audit \
	dnsZoneId=/subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.Network/dnszones/xxx \
	azureKeyvaultSecretsProvider=true \
	createKV=true \
	kvOfficerRolePrincipalId=$(az ad signed-in-user show --query id --out tsv)
```

khowling commented on bicep/main.bicep:
```diff
@@ -171,21 +170,17 @@ var appGwSubnetId = ingressApplicationGateway ? (custom_vnet ? network.outputs.a
-  scope: resourceGroup(dnsZoneRg)
   params: {
-    dnsZoneName: dnsZoneName
-    isPrivate: isPrivate
```
@Gordonby i think there are a bunch of issues here, you're not passing `dnsZoneId` or any parts of it to the module

Gordonby replied:
Lol. Good spot.
For some reason the module allowed that param to be optional.

All fixed up.

khowling commented on bicep/main.bicep:
```diff
@@ -171,21 +170,17 @@ var appGwSubnetId = ingressApplicationGateway ? (custom_vnet ? network.outputs.a
 
-module dnsZone './dnsZone.bicep' = if (!empty(dnsZoneId)) {
+module dnsZone './dnsZoneRbac.bicep' = if (!empty(dnsZoneId)) {
   name: 'addDnsContributor'
```
you have this deployment name in 2 places, better to make them distinct so can track it in the portal deployment panel

Gordonby commented:
I've addressed your points @khowling 

What was really missing here was the BYO vnet CI/CD doing a full run. It leverages a dns name for it's application. Had it run, i would have spotted these issues earlier.  I've made a change to its CI/CD trigger so that it does a full run if the `test-deploy-byoconfig` label is on the PR

Manual test of it here:
https://github.com/Azure/AKS-Construction/runs/7404444607?check_suite_focus=true
![image](https://user-images.githubusercontent.com/17914476/179700318-34cfe8e2-1fb0-4a52-a6e7-3c38b4b0c243.png)

Can you re-run your test again please?
